### PR TITLE
Fix: download leveldbstore plugin.

### PR DESCRIPTION
**Pull request by Jim8y**
this pr fix the issue that we can not directly run the neo-cli because of lacking the leveldb lib and leveldbstore.

it:

- load the leveldb according to the system
- download leveldbstore if user agree.

leveldb libs are from the release version of the neo-node made by @superboyiii. 
But i compiled the version for mac os arm on my own machine cause @superboyiii  you put the wrong leveldb for the macarm.

### CODE REVIEW

**Comment by Liaojinghui** [comment]
with this pr, you can run the neo-cli directly.

File: src/Neo.CLI/CLI/MainService.Plugins.cs [hunk 1]
```diff
@@ -241,5 +241,58 @@ private void OnPluginsCommand()
...
+                    output.Write(buffer, 0, read);
+                    totalRead += read;
+                    Console.Write(
+                        $"\rDownloading {pluginName}.zip {totalRead / 1024}KB/{response.Content.Headers.ContentLength / 1024}KB {(totalRead * 100) / response.Content.Headers.ContentLength}%");
+                }
+
+                Console.WriteLine();
+                return output;
+            }
+        }
```
Comment (Liaojinghui) [inline src/Neo.CLI/CLI/MainService.Plugins.cs:296]:
```suggestion
```

**Comment by Liaojinghui** [comment]
Screenshot 2023-12-14 at 12 21 16 AM

**Comment by cschuchardt88** [comment]
Please see #3018

**Comment by cschuchardt88** [comment]
We do not want binaries in the repo, please. Also i can update the workflow to fix this to release the binaries for you.

**Comment by Liaojinghui** [comment]
> We do not want binaries in the repo, please. Also i can update the workflow to fix this to release, the binaries for you. Plus we are working on apt-get and brew repos for this project, so binary package will be installed automatically.

This is not for release. This is for build and test. You can still have your standalone workflow, but that will not solve the issue this pr trys to address.

This pr does not necessarily to be the final solution, but a proper alternative solution should be proposed to replace this one. For instance, a script to setup the build and run environment.

**Comment by Liaojinghui** [comment]
@cschuchardt88 Let me explain the issue here. When we clone the core and build it, everything would work as expected, but when we try to run it, there are two issues, the first is that we dont have a leveldb in the project, we need to install it manually, (this sucks, i would expect run a script to setup the environment, or one click to run), and another issue is that it must rely on leveldbstore plugin to run, while it requires to run the cli to install the leveldbstore plugin, its a deadlock.

**Comment by roman-khimov** [comment]
These are not small binaries, adding them into the repo is no good.

**Comment by Liaojinghui** [comment]
> These are not small binaries, adding them into the repo is no good.

This not necessarily to be the final solution, binary is not good, i am also seeking to download from neo-project repo, just like plugins. Again, this is an issue, we need to fix it, please share your solution.

**Comment by Liaojinghui** [comment]
Let this pr be more focused on downloading the leveldbstore plugin then

File: src/Neo.CLI/CLI/MainService.cs [hunk 2]
```diff
@@ -605,5 +611,11 @@ static string GetExceptionMessage(Exception exception)
 
             return exception.Message;
         }
+
+        private Task TryInstallLevelDbPlugin()
+        {
+            if (PluginExists("LevelDBStore")) return Task.CompletedTask;
+            return !ReadUserInput("LevelDBStore plugin is required but not installed. Do you want to install it now? (yes/no)").IsYes() ? Task.CompletedTask : InstallPluginAsync("LevelDBStore");
```
Comment (lock9) [inline src/Neo.CLI/CLI/MainService.cs:618]: I can see that you are reusing code to install plugins, but it's not working for some reason:

```
Unhandled exception. System.Collections.Generic.KeyNotFoundException: The given key 'LevelDBStore' was not present in the dictionary.
   at System.Collections.Generic.Dictionary`2.get_Item(TKey key)
   at Neo.Persistence.StoreFactory.GetStore(String storageEngine, String path) in C:\Users\Windows11\Workspace\monorepo\neo\src\Neo\Persistence\StoreFactory.cs:line 37
...
```

File: src/Neo.CLI/CLI/MainService.cs [hunk 3]
```diff
@@ -356,6 +360,8 @@ public async void Start(string[] args)
                         break;
                 }
 
+            await TryInstallLevelDbPlugin();
```
Comment (lock9) [inline src/Neo.CLI/CLI/MainService.cs:363]: It's asking if I want to install LevelDB even if I'm using Memory Store

Comment (Liaojinghui) [reply src/Neo.CLI/CLI/MainService.cs:369]: fixed.

File: src/Neo.CLI/CLI/MainService.cs [hunk 4]
```diff
@@ -66,6 +68,8 @@ public NeoSystem NeoSystem
 
         protected override string Prompt => "neo";
         public override string ServiceName => "NEO-CLI";
+        static readonly string StrExeFilePath = Path.GetDirectoryName(Assembly.GetExecutingAssembly().Location);
+        static readonly string PluginPath = Path.Combine(StrExeFilePath, "Plugins");
```
[the remark this answers is not in the document]
Comment (Liaojinghui) [reply src/Neo.CLI/CLI/MainService.cs]: fixed

File: scripts/install-requirements.sh [hunk 5]
```diff
@@ -0,0 +1,57 @@
+#!/bin/bash
+
+# Detect the operating system type
+OS="$(uname -s)"
+DISTRO=""
+
+# Define installation functions
+install_leveldb_ubuntu() {
+    echo "Installing LevelDB on Ubuntu..."
+    sudo apt-get update && sudo apt-get install -y libleveldb-dev
```
Comment (shargon) [inline scripts/install-requirements.sh:10]: This file is for leveldb, but maybe the user want to use RocksDb

Comment (Liaojinghui) [reply scripts/install-requirements.sh:10]: This is the tricky part, do we have any instruction of saying to install rockdb? i dont even know we need to install it.

Comment (cschuchardt88) [reply scripts/install-requirements.sh:10]: what about `OpenSuSE` linux? That's what i use.
```sh
zypper ref && zypper in leveldb-devel
```

File: scripts/install-requirements.sh [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+}
+
+install_requirements_fedora() {
+    echo "Installing requirements and dependencies on Fedora..."
+    sudo dnf install -y leveldb-devel sqlite sqlite-devel libunwind-devel
+}
+
+install_requirements_opensuse() {
+    echo "Installing requirements and dependencies on openSUSE..."
+    sudo zypper install -y libleveldb-devel sqlite3 libsqlite3-devel libunwind-devel
```
Comment (cschuchardt88) [inline scripts/install-requirements.sh:21]: Don't forget to add refresh **no** `-y`
```sh
zypper refresh
```

Also correct command is
```sh
zypper install -y leveldb-devel sqlite3-devel
```

Complete command is:
```sh
zypper refresh && zypper install -y leveldb-devel sqlite3-devel
```

File: src/Neo.CLI/CLI/MainService.cs [hunk 7]
```diff
@@ -323,16 +324,26 @@ private byte[] LoadUpdateScript(UInt160 scriptHash, string nefFilePath, string m
             }
         }
 
-        public override void OnStart(string[] args)
+        public override async Task OnStart(string[] args)
```
Comment (cschuchardt88) [inline src/Neo.CLI/CLI/MainService.cs:327]:
```suggestion
        public override async Task OnStartAsync(string[] args)
```

File: src/Neo.CLI/CLI/MainService.cs [hunk 8]
```diff
@@ -624,5 +635,41 @@ private Task TryInstallLevelDbPlugin()
             if (PluginExists(Settings.Default.Storage.Engine)) return Task.CompletedTask;
             return !ReadUserInput($"{Settings.Default.Storage.Engine} plugin is required but not installed. Do you want to install it now? (yes/no)").IsYes() ? Task.CompletedTask : InstallPluginAsync(Settings.Default.Storage.Engine);
         }
+
+        private async Task TryDownloadLevelDbDll()
```
Comment (cschuchardt88) [inline src/Neo.CLI/CLI/MainService.cs:639]:
```suggestion
        private async Task TryDownloadLevelDbDllAsync()
```

File: src/Neo.ConsoleService/ConsoleServiceBase.cs [hunk 9]
```diff
@@ -281,12 +281,13 @@ protected void OnExit()
 
         #endregion
 
-        public virtual void OnStart(string[] args)
+        public virtual Task OnStart(string[] args)
```
Comment (cschuchardt88) [inline src/Neo.ConsoleService/ConsoleServiceBase.cs:284]:
```suggestion
        public virtual Task OnStartAsync(string[] args)
```

File: src/Neo.CLI/CLI/MainService.cs [hunk 10]
```diff
@@ -323,16 +324,26 @@ private byte[] LoadUpdateScript(UInt160 scriptHash, string nefFilePath, string m
             }
         }
 
-        public override void OnStart(string[] args)
+        public override async Task OnStart(string[] args)
         {
             try
             {
-                base.OnStart(args);
+                await base.OnStart(args);
```
Comment (cschuchardt88) [inline src/Neo.CLI/CLI/MainService.cs:331]:
```suggestion
                await base.OnStartAsync(args);
```

**Review by cschuchardt88 — CHANGES_REQUESTED** [review]
If your going to do `async` please follow naming convention for method names.

**Comment by Liaojinghui** [comment]
Mind to directly give suggestions to the code?

File: scripts/install-requirements.sh (hunk 6 above)
Comment (cschuchardt88) [inline scripts/install-requirements.sh:21]:
```suggestion
    sudo zypper refresh
    sudo zypper install -y leveldb-devel sqlite3-devel
```

**Review by cschuchardt88 — CHANGES_REQUESTED** [review]
What do you mean? It should be on `Files changed` tab?

**Comment by cschuchardt88** [comment]
Youa need to add me as `Contributor` to your repo if want me to do that.
```
Remote: Permission to Liaojinghui/neo.git denied to cschuchardt88.`
```

File: src/Neo.ConsoleService/ServiceProxy.cs [hunk 11]
```diff
@@ -23,7 +23,7 @@ public ServiceProxy(ConsoleServiceBase service)
 
         protected override void OnStart(string[] args)
         {
-            _service.OnStart(args);
+            _service.OnStartAsync(args);
```
Comment (shargon) [inline src/Neo.ConsoleService/ServiceProxy.cs:26]: This must be awaited

File: src/Neo.CLI/Neo.CLI.csproj [hunk 12]
```diff
@@ -25,5 +25,4 @@
     <ProjectReference Include="..\Neo.ConsoleService\Neo.ConsoleService.csproj" />
     <ProjectReference Include="..\Neo\Neo.csproj" />
   </ItemGroup>
-
 </Project>
```
Comment (shargon) [inline src/Neo.CLI/Neo.CLI.csproj:28]: Try to don't modify files that are not related with the PR

**Comment by Jim8y** [comment]
@shargon @vncoelho @cschuchardt88 please take a look, would be great to have it merged, when i debug or work, always need to handle the leveldbplugin and leveldb.dll manually.

**Comment by vncoelho** [comment]
Not sure about this solution.

Currently it is working by running just with the local memory? 

Let me check with carefull and verify the case.

**Comment by Jim8y** [comment]
> Not sure about this solution.
> 
> Currently, is it working by running just with the local memory?
> 
> Let me check with carefull and verify the case.

It leveldb by default.

**Comment by vncoelho** [comment]
Strange, from what I remember the default when we launched N3 would be local memory.

**Comment by cschuchardt88** [comment]
Maybe we should switch to `RocksDb`. Would make resync easier. I have a solution in mind. However `RocksDb` doesn't have 32-bit version.

File: src/Neo.CLI/CLI/MainService.cs [hunk 13]
```diff
@@ -636,5 +659,48 @@ static string GetExceptionMessage(Exception exception)
...
+        private Task TryInstallLevelDbPluginAsync()
+        {
+            if (Settings.Default.Storage.Engine != "LevelDBStore" && Settings.Default.Storage.Engine != "RocksDBStore") return Task.CompletedTask;
+            if (PluginExists(Settings.Default.Storage.Engine)) return Task.CompletedTask;
+            return !ReadUserInput($"{Settings.Default.Storage.Engine} plugin is required but not installed. Do you want to install it now? (yes/no)").IsYes() ? Task.CompletedTask : InstallPluginAsync(Settings.Default.Storage.Engine);
+        }
+
+        private async Task TryDownloadLevelDbDllAsync()
+        {
+            string url = "https://github.com/neo-ngd/leveldb/releases/download/v1.23/libleveldb-win-x64.zip";
```
Comment (cschuchardt88) [inline src/Neo.CLI/CLI/MainService.cs:672]: Do not forget about `win-arm64`. #3018 does build arm64 for `neo-cli`.

**Review by shargon — COMMENTED** [review]
LGTM if it was tested

**Comment by superboyiii** [comment]
You forgot centOS

**Comment by cschuchardt88** [comment]
@Jim8y Dont worry about this this, i have it all in #3018, I can deploy on release if need, just put in comment on that PR.

@superboyiii remember what you told me, about `libleveldb-dev` not being current. Well same problem here.

**Comment by Jim8y** [comment]
this is for local build, not workflow.

**Comment by cschuchardt88** [comment]
@Jim8y What i mean is, i can build all the current `libleveldb-dev` libraries for all OSs & CPUs in that PR and release on this repo for you to download in `neo-cli`. So we alway have the current `leveldb` version. Or else we will run into this problem. https://github.com/neo-project/neo/pull/3018#issuecomment-1876875462

**Comment by Jim8y** [comment]
@shargon

**Comment by shargon** [comment]
> LGTM if it was tested

@superboyiii tested?

File: scripts/install-requirements.sh [hunk 14]
```diff
@@ -0,0 +1,67 @@
+#!/bin/bash
```
Comment (shargon) [inline scripts/install-requirements.sh:1]: Maybe we can include this script as dependency-installer script in the plugin, and if this file is present execute it during the installation?

Comment (cschuchardt88) [reply scripts/install-requirements.sh:1]: Or we can use #3018 and problem solved

Comment (shargon) [reply scripts/install-requirements.sh:1]: > Or we can use #3018 and problem solved

Different problems, this is for execute a script installation for plugin

Comment (cschuchardt88) [reply scripts/install-requirements.sh:1]: Yes but with deb packages it auto installs all those dependencies on linux.

**Review by shargon — CHANGES_REQUESTED** [review]
@Jim8y I prefer to move the script to the specific module, and execute here if it's present

**Comment by Jim8y** [comment]
> @Jim8y I prefer to move the script to the specific module, and execute here if it's present

We will have many other scripts for the ease of setup and run neo node, they all should be kept together. This is just the begining. If we keep scripts here and there, it will be hard to find.

**Comment by shargon** [comment]
> > @Jim8y I prefer to move the script to the specific module, and execute here if it's present
> 
> We will have many other scripts for the ease of setup and run neo node, they all should be kept together. This is just the begining. If we keep scripts here and there, it will be hard to find.

Not here, here we can detect if it's present, each module will have his own scripts

**Comment by Jim8y** [comment]
scripts are tp helping users, name amd readme is enough to tell them how to use, if you put them here and there, hard to follow. again check some other projects to see where are scripts, neo-go for example.

**Comment by shargon** [comment]
> scripts are tp helping users, name amd readme is enough to tell them how to use, if you put them here and there, hard to follow. again check some other projects to see where are scripts, neo-go for example.

This script is module dependant, is not the same

**Comment by Jim8y** [comment]
Scripts are used by users, not the neo modules. User need scripts, they go scripts, we dont hide scripts inside of the project here and there.

**Comment by shargon** [comment]
> Scripts are used by users, not the neo modules. User need scripts, they go scripts, we dont hide scripts inside of the project here and there.

this script is used by the client, no by the user, it's transparent to the user, and prepare the specific plugin installation

**Comment by Jim8y** [comment]
> this script is used by the client, no by the user, it's transparent to the user, and prepare the specific plugin installation

we dont use client to run any script transparent, it has security concerns. User ran script, client does not.

File: src/Neo.CLI/CLI/MainService.cs [hunk 15]
```diff
@@ -345,11 +347,11 @@ private byte[] LoadUpdateScript(UInt160 scriptHash, string nefFilePath, string m
...
         }
 
-        public override void OnStart(string[] args)
+        public override Task OnStartAsync(string[] args)
         {
-            base.OnStart(args);
+            base.OnStartAsync(args).GetAwaiter().GetResult();
             OnStartWithCommandLine(args);
-
+            return Task.CompletedTask;
```
Comment (cschuchardt88) [inline src/Neo.CLI/CLI/MainService.cs:354]: use `await` instead. If since you are getting the `result`.

**Review by cschuchardt88 — CHANGES_REQUESTED** [review]

**Review by shargon — COMMENTED** [review]
@Jim8y I really think that this is useful, but designed in a different way, the script must come with the plugin, and if the plugin zip contains the script, execute it before install the plugin, in this way it could be used for others plugins, like rocksDb for example.

- PR1: move the script to neo-modules
- PR2: here check if the script exists and execute it during installation.

**Comment by Jim8y** [comment]
Just go ahead do what you think  is the best. I am not going to against you.
